src/appleseed/renderer/modeling/postprocessingstage/vignettepostprocessingstage.cpp

```diff
@@ -0,0 +1,196 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "vignettepostprocessingstage.h"
+
+// appleseed.renderer headers.
+#include "renderer/modeling/frame/frame.h"
+#include "renderer/modeling/postprocessingstage/postprocessingstage.h"
+
+// appleseed.foundation headers.
+#include "foundation/containers/dictionary.h"
+#include "foundation/image/canvasproperties.h"
+#include "foundation/image/color.h"
+#include "foundation/image/image.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+
+using namespace foundation;
+
+namespace renderer
+{
+
+namespace
+{
+    //
+    // Vignette post-processing stage.
+    //
+
+    const char* Model = "vignette_post_processing_stage";
+
+    const float DefaultIntensity = 0.5f;
+    const float MinIntensity = 0.0f;
+    const float MaxIntensity = 1.0f;
+
+    class VignettePostProcessingStage
+        : public PostProcessingStage
+    {
+      public:
+        VignettePostProcessingStage(
+            const char*             name,
+            const ParamArray&       params)
+          : PostProcessingStage(name, params)
+        {
+        }
+
+        void release() override
+        {
+            delete this;
+        }
+
+        const char* get_model() const override {
+            return Model;
+        }
+
+        bool on_frame_begin(
+            const Project&          project,
+            const BaseGroup*        parent,
+            OnFrameBeginRecorder&   recorder,
+            IAbortSwitch*           abort_switch) override
+        {
+            const OnFrameBeginMessageContext context("post-processing stage", this);
+
+            m_intensity = m_params.get_optional("intensity", DefaultIntensity, context);
+
+            return true;
+        }
+
+        void execute(Frame& frame) const override
+        {
+            const CanvasProperties& props = frame.image().properties();
+            const Vector2f resolution(static_cast<float>(props.m_canvas_width), static_cast<float>(props.m_canvas_height));
+            const Vector2f aspect(resolution.x / resolution.y, 1);
+
+            Image& image = frame.image();
+
+            for (size_t y = 0; y <= resolution.y - 1; ++y)
+            {
+                for (size_t x = 0; x <= resolution.x - 1; ++x)
+                {
+                    Vector2f coord = (2.0f * Vector2f(static_cast<float>(x), static_cast<float>(y)) - resolution) / resolution.y;
+
+                    //
+                    // Port of Keijiro Takahashi's natural vignetting effect for Unity.
+                    //
+                    // Reference:
+                    //
+                    //   https://github.com/keijiro/KinoVignette
+                    //
+
+                    float rf = norm(coord) * m_intensity;
+                    float rf2_1 = rf * rf + 1.0f;
+                    float e = 1.0f / (rf2_1 * rf2_1);
```

Comment:
Can you find better names for the variables and maybe comment it ?

Reply:
I asume `rf` probably means "radial falloff", since this is what it computes, but I'm not really sure about `e`. Also, since these three lines are a direct port of the [original implementation](https://github.com/keijiro/KinoVignette/blob/master/Assets/Kino/Vignette/Shader/Vignette.shader#L41) (which is not commented), I thought it'd be better not to rename them.

What do you think about leaving the names as they are, but commenting next to `e` that it's " inversely proportional to the fourth power of the distance to the center"? 😅

Reply:
>  I thought it'd be better not to rename them

Why ? 

If you rename `rf` to `radial_falloff`, the code becomes much clearer. And `e` to `inverse_radial_falloff`.

Reply:
> Why ?

I couldn't think of a good name for `e`, but `inverse_radial_falloff` or `inverse_quartic_radial_falloff` is nice (and makes the comment redundant so we can remove it). I'll change it 🙂